Source sampling executable

DanShort12 opened the pull request:
Adds an executable called `source_generator` to allow an `initial_source.h5` to be created without having to run a full OpenMC simulation.

This approach utilises the same OpenMC commands that would be run when running within the simulation. In particular:

*  `calculate_work`
*  `allocate_banks`
*  `initialize_source`

It also controls the settings needed to generate the initial source, allowing the user to specify the number of particles to sample, the output path, the verbosity, and the location of the source library (e.g. as built by the parametric plasma source). The run mode is fixed to eigenvalue, as this is the only OpenMC mode that generates the initial source output file, and the write initial source setting is fixed at true so that the output is always generated.

I've also included updates to the README to describe the build and run process.

Commit:
made python driver for the source

Commit:
changed version number

Commit:
returned from $keywords$ to default exisiting code

Commit:
added MIT lic

Commit:
removed script folder

Commit:
moved cpp files into python to solve not finding file error

shimwell committed:
Update README.md

Commit:
removed print statement

Commit:
Merge branch 'python_version' of github.com:Shimwell/parametric-plasma-source into python_version

Commit:
removed minor radius limit

DanShort12 committed:
Create a source_generator and build it

The source_generator provides a means to execute the source
sampling outside of an OpenMC simulation. This may be of interest
for diagnostic and testing purposes.

Add the relevant information to the Makefile to build the
source_generator and clean resulting binaries.

Add items to .gitignore so that executable and output are not
picked up.

DanShort12 committed:
Create a source_generator and build it

The source_generator provides a means to execute the source
sampling outside of an OpenMC simulation. This may be of interest
for diagnostic and testing purposes.

Add the relevant information to the Makefile to build the
source_generator and clean resulting binaries.

Add items to .gitignore so that executable and output are not
picked up.

DanShort12 committed:
Create a source_generator and build it

The source_generator provides a means to execute the source
sampling outside of an OpenMC simulation. This may be of interest
for diagnostic and testing purposes.

Add the relevant information to the Makefile to build the
source_generator and clean resulting binaries.

Add items to .gitignore so that executable and output are not
picked up.

DanShort12 committed:
Add in lines from .gitignore missed in last commit

DanShort12 committed:
Add in lines from .gitignore missed in last commit

DanShort12 committed:
Add in lines from .gitignore missed in last commit

DanShort12 committed:
Updated the README with build/run instructions

DanShort12 committed:
Updated the README with build/run instructions

DanShort12 committed:
Updated the README with build/run instructions

DanShort12 committed:
Add the temperature beta value

The Fausser paper describes the exponent on the ratio of the sampled
radius to the pedestal radius as being a variable beta_T, but the
current implementation fixes this at a value of 2. This commit makes
the beta_T value variable and sets the default value of 6.0 as used in
the paper.

DanShort12 committed:
Remove unused sig_dt variable

DanShort12 committed:
Scale pedistal radius value by minor radius

The Fausser paper gives a_ped as 0.8 * A but the code currently just
uses the dimensionless value of 0.8. Scale this up by the minor radius
in the default input values to give the pedestal radius in metres.

DanShort12 committed:
Read C++ and Makefile files into Python file

This commit provides a utility to read the C++ and Makefile files into
a Python file so that the code can be updated via a script when a new
C++ version is available. This is a halfway house to automatically
generating the C++, as it is not possible to create a shared object
library directly from Python.

Also updates .gitignore to ignore compiled Python files.

DanShort12 committed:
Avoid matching on variable values

The default values in the C++ version may change over time, so
it is best to avoid matching on those to ensure that the Python
version doesn't need fine-grain updates when the C++ changes.

DanShort12 committed:
Add pybind as a submodule

DanShort12 committed:
CMakeLists for original build configuration

Mimics functionality of existing Makefile.

Ignore build directory in git.

DanShort12 committed:
Create Python bindings and build

Add a pybind.cpp file and use it to wrap the basic plasma source
functionality.

Update CMakeLists to build the Python bindings.

Ignore untracked Python artefacts in git.

DanShort12 committed:
Fix indentation

DanShort12 committed:
Install bindings with setup.py

Include instructions in setup.py to compile the bindings with cmake.
Specifically only for the plasma_source library, since we don't need
the source_sampling library in Python.

Update CMakeLists to avoid forcing the library output directory as this
conflicts with the way setuptools etc. look for package artefacts. This
means that the resulting source_sampling library is available in the
build directory, unless the CMAKE_LIBRARY_OUTPUT_DIRECTORY is
passed into the cmake configuration step.

Introduce an __init__.py so that the modules can easily be loaded and
found by setuptools.

Cleanup .gitignore and add dist and egg paths.

makeclean committed:
Merge pull request #5 from DanShort12/scale_pedestal_radius

Scale pedestal radius value by minor radius

makeclean committed:
Merge pull request #4 from DanShort12/add_temperature_beta

Add temperature beta and remove sig_dt

DanShort12 committed:
Merge pull request #1 from makeclean/master

Pick up changed in upstream branch

shimwell committed:
Merge pull request #2 from DanShort12/cmake_pybind

Python bindings

shimwell committed:
Merge branch 'python_version' into read_cpp_files

shimwell committed:
Merge pull request #1 from DanShort12/read_cpp_files

Read C++ and Makefile files into Python file

DanShort12 committed:
Merge branch 'master_upstream' into pickup_prs_from_makeclean

DanShort12 committed:
Update source

DanShort12 committed:
Tidy up gitignore

DanShort12 committed:
Add bindings for temperature beta

DanShort12 committed:
Fix source import error

DanShort12 committed:
Add temperature beta to Python

DanShort12 committed:
Pedestal radius should be in length units

The pedestal radius used should have dimensionality of the length units
being used.

DanShort12 committed:
Build python in setup

Ensure that the Python build method is called when running setup so that
the C++ template strings are always in line with the C++ code.

shimwell committed:
Merge pull request #4 from DanShort12/build_python_in_setup

Build python in setup

shimwell committed:
Merge pull request #3 from DanShort12/pickup_prs_from_makeclean

Pickup PRs from makeclean

makeclean reviewed (commented):
Can you cmakeify that makefile?

shimwell committed:
fixed url for repo

shimwell committed:
sorry copy paste error

DanShort12 committed:
Add plugixml as a submodule

This aligns with XML handling used by OpenMC.

DanShort12 committed:
Get submodules when CMake-ing

DanShort12 committed:
Template serialisation routines

Introduces a to_xml serialisation routine that can write to string or to
file.
Creates the Python bindings for that routine.

DanShort12 committed:
Serialise to XML

Adds functionality to serialise a PlasmaSource to XML either as a string
or to a file.
Implements XML handling using pugixml and uses a hanlder class for XML
operations.

DanShort12 committed:
Example serialised representation.

DanShort12 committed:
Deserialise the plasma source from XML

Deserialises the plasma source from an XML file and create an example
input file.

DanShort12 committed:
Fix example source

DanShort12 committed:
Deserialise from an XML string

DanShort12 committed:
Remove duplicate code and simplify

Moves duplicate code into private function and passed XML doc as
pointer.
Adds helper functions for getters.
Moves root name to const.

DanShort12 committed:
Use root name consistently

DanShort12 committed:
Specify max toroidal angle in degrees

DanShort12 committed:
Serialise angles in degrees

makeclean committed:
Merge pull request #2 from Shimwell/python_version

Python version

makeclean commented:
Sorry @DanShort12 can you rebase, then ill get this merged.

DanShort12 committed:
Merge branch 'source_sampling_executable' of https://github.com/DanShort12/parametric-plasma-source into source_sampling_executable

DanShort12 committed:
Pick up changes from makeclean

DanShort12 committed:
CMake configuration for source_generator

DanShort12 commented:
Rebased and cmakified.

DanShort12 committed:
First pass at python CI config

DanShort12 committed:
Use empty init file

This prevents an error on a fresh build

DanShort12 committed:
Build using bdist_wheel rather than sdist

DanShort12 committed:
Make sure wheel is installed

DanShort12 committed:
Get submodules in CI

shimwell commented on README.md:
```diff
@@ -1,5 +1,68 @@
+from parametric_plasma_source import Plasma
+my_plasma = Plasma(major_radius=6,
+                   minor_radius=1.5,
+                   elongation = 2.0
```
```suggestion
                   elongation = 2.0,
```

DanShort12 replied:
Thanks @Shimwell - I've fixed the typo and pushed to this PR.

shimwell commented:
Does anyone else get this error when trying ```python setup.py install```

```
  File "setup.py", line 8, in <module>
    from parametric_plasma_source.build_python import build as build_python
  File "/home/jshim/parametric-plasma-source/parametric_plasma_source/__init__.py", line 2, in <module>
    from .plasma_source import PlasmaSource
ModuleNotFoundError: No module named 'parametric_plasma_source.plasma_source'
```

DanShort12 committed:
Only build plasma_source for Python

DanShort12 committed:
Clear init file

Do not collect imports as the package might not be built, in which case
the imports could fail.

DanShort12 committed:
Fix typo in README

shimwell commented:
That appeared to help somewhat but not quite fully there with the setup.py, ```python setup.py install``` latest error ...
```
running develop
running egg_info
writing parametric_plasma_source.egg-info/PKG-INFO
writing dependency_links to parametric_plasma_source.egg-info/dependency_links.txt
writing top-level names to parametric_plasma_source.egg-info/top_level.txt
reading manifest file 'parametric_plasma_source.egg-info/SOURCES.txt'
writing manifest file 'parametric_plasma_source.egg-info/SOURCES.txt'
running build_ext
CMake Error at CMakeLists.txt:43 (add_subdirectory):
  The source directory

    /home/jshim/parametric-plasma-source/pybind11

  does not contain a CMakeLists.txt file.


CMake Error at CMakeLists.txt:45 (pybind11_add_module):
  Unknown CMake command "pybind11_add_module".


-- Configuring incomplete, errors occurred!
See also "/home/jshim/parametric-plasma-source/build/temp.linux-x86_64-3.7/CMakeFiles/CMakeOutput.log".
Traceback (most recent call last):
  File "setup.py", line 84, in <module>
    "Operating System :: OS Independent",
  File "/home/jshim/anaconda3/lib/python3.7/site-packages/setuptools/__init__.py", line 144, in setup
    return distutils.core.setup(**attrs)
  File "/home/jshim/anaconda3/lib/python3.7/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/home/jshim/anaconda3/lib/python3.7/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/home/jshim/anaconda3/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/home/jshim/anaconda3/lib/python3.7/site-packages/setuptools/command/develop.py", line 38, in run
    self.install_for_development()
  File "/home/jshim/anaconda3/lib/python3.7/site-packages/setuptools/command/develop.py", line 140, in install_for_development
    self.run_command('build_ext')
  File "/home/jshim/anaconda3/lib/python3.7/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/home/jshim/anaconda3/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "setup.py", line 26, in run
    self.build_extension(ext)
  File "setup.py", line 56, in build_extension
    env=env
  File "/home/jshim/anaconda3/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '/home/jshim/parametric-plasma-source', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/home/jshim/parametric-plasma-source/parametric_plasma_source/', '-DPYTHON_EXECUTABLE=/home/jshim/anaconda3/bin/python', '-DCMAKE_BUILD_TYPE=Release']' returned non-zero exit status 1.
```

DanShort12 committed:
Get submodules in CMake configuration

This ensures that users who try to build the package have the source
for the submodules available.

DanShort12 commented:
Couple of issues with building on a clean environment... all the more reason to get CI working! Thanks for trying it out @Shimwell - can you try again when you get a mo and let me know if it's working now?

shimwell commented:
oh wow it looks like it works, that is a fancy setup.py it clones  git repos automatically and builds things.

shimwell commented:
I shall continue to test, I'm not sure if this will need some more changes to get it working on the pypi repo but let me tinker for the day

DanShort12 commented:
Pretty sure it is likely to need a bit of tinkering for the PyPI repo but appreciate if you get a bit of time to try it out IRL (my poking has mostly been theoretical). I can rummage to try to find my script to read the initial_source.h5 into matplotlib scatter plots if that might help.

DanShort12 commented:
Oh, and the setup.py magic is in the interplay between the build and the CMakeLists.txt file - if you fancy a look through that file. The last bit of tinkering was to add some git commands into that CMake file.

Since the bindings are compiled, we'll need a bit of a think about how to package them for PyPI... they will be compile and Python version dependent.

DanShort12 committed:
Add plasma module back into __init__.py

This avoids a regression effect against the documentation in README.md.

DanShort12 committed:
Fix order of build args

Target must come before the make arguements.

DanShort12 committed:
Do not fail if openmc not found

Makes OpenMC optional but does not build the source_sampling library if
OPENMC_LIB is not found.

shimwell commented:
Perhaps this is just me but the ./source_generator --output is behaving slightly strangely

This command
``` ./source_generator -l test.so -o test.h5```
prints this to the terminal
```
 Number of particles: 1000
 Source library: test.so
 Output path: test.h5
 Verbosity: 5
```
but then makes a .h5 file called test.h5initial_source.h5

So it might be appending `initial_source.h5` to every output filename

DanShort12 committed:
Remove unused RANDOM definition

Random numbers are only generated in the source sample_source method, so
we shouldn't need to link against OpenMC where we don't use OpenMC
functions.

DanShort12 committed:
Upload wheel as artefact

DanShort12 committed:
Create new publish step to upload wheels

DanShort12 committed:
Tidy up actions

DanShort12 committed:
Fix syntax error

DanShort12 committed:
Another syntax fix

DanShort12 committed:
Revert workflow to single job

DanShort12 committed:
Implement a simple test

DanShort12 committed:
Don't run compile test if OpenMC not installed

DanShort12 committed:
Add some more tests

DanShort12 committed:
Add status badge

Also a few format update to README.

DanShort12 committed:
Use a clearer job name

DanShort12 committed:
Only run actions when affecting master

DanShort12 committed:
Merge pull request #2 from makeclean/master

Merge makeclean master into fork

DanShort12 committed:
Submodules are now retrieved from CMake config

DanShort12 committed:
Pick up submodules from CMake config

DanShort12 committed:
Tidy up step names

shimwell commented:
I've been tinkering and made some plots to show the plasma source with a few parameters chaining. I only tried the 4 shaping parameters (elongation, major_radius, minor_radius, triangularity) at the moment. I got stuck on shafranov_shift which appeared to have no impact on the plasma. But I need to check I'm doing everything correctly to be certain.


![minor_radius](https://user-images.githubusercontent.com/8583900/88947809-fa99f380-d288-11ea-876a-735a4036b045.gif) 
![triangularity](https://user-images.githubusercontent.com/8583900/88947821-038ac500-d289-11ea-8614-7d3ee05f7130.gif)


![major_radius](https://user-images.githubusercontent.com/8583900/88947788-f1a92200-d288-11ea-889d-1a1bfe968b65.gif)

![elongation](https://user-images.githubusercontent.com/8583900/88947769-ea821400-d288-11ea-84a3-39fb8a98de07.gif)

DanShort12 commented:
Thanks @Shimwell, looks great! Just a suggestion, but the plots might be easier to interpret if they have some text in the corner to show what variable is changing and what the value being used is for a given frame (I know it's possible to deduce the variable from how the plot changes but that requires a little bit of bonus thought).

makeclean commented:
Hmm shafranov shift should shift the equator radially outwards IIRC. Looking at the code as long as you’re using the sample source function it should work fine, if you short cut it and go some other route that avoids convert_rad_to_rz then you’ll not get the shift.

On 30 Jul 2020, at 17:23, Jonathan Shimwell <notifications@github.com<mailto:notifications@github.com>> wrote:



I've been tinkering and made some plots to show the plasma source with a few parameters chaining. I only tried the 4 shaping parameters (elongation, major_radius, minor_radius, triangularity) at the moment. I got stuck on shafranov_shift which appeared to have no impact on the plasma. But I need to check I'm doing everything correctly to be certain.

[minor_radius]<https://user-images.githubusercontent.com/8583900/88947809-fa99f380-d288-11ea-876a-735a4036b045.gif>
[triangularity]<https://user-images.githubusercontent.com/8583900/88947821-038ac500-d289-11ea-8614-7d3ee05f7130.gif>

[major_radius]<https://user-images.githubusercontent.com/8583900/88947788-f1a92200-d288-11ea-889d-1a1bfe968b65.gif>

[elongation]<https://user-images.githubusercontent.com/8583900/88947769-ea821400-d288-11ea-84a3-39fb8a98de07.gif>

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/makeclean/parametric-plasma-source/pull/3#issuecomment-666504346>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AASTUSXSQDKIPJUEU54R62TR6GNADANCNFSM4OCRWMFQ>.

DanShort12 committed:
Updates to work with OpenMC deserialisation

Align with work in progress on OpenMC to support deserialisation of a
serialised source. In particular remove XMLHelper in deserialisation as
this causes some symbols to be unfindable.
Also removes default values and object from sampling as no longer
needed.

DanShort12 committed:
Include source sampling in bindings

Handles pass by reference using a lambda that maps return values to a
tuple.
Also tweak some doc strings.

DanShort12 committed:
Fix typo

DanShort12 committed:
Strip out build files that are no longer needed

As the functionality will be used via serialisation, this means that we
won't need to recompile the C++ library.

DanShort12 committed:
Update serialization methods

Aligns with recent version of OpenMC:
*  Serialize to key-value pair string;
*  Expose deserialization via openmc_create_source
*  Remove XML handling
*  Remove pugixml as a submodule

Some formatting updates

DanShort12 committed:
Update to align with OpenMC changes

Use unique_ptr when creating source.

DanShort12 committed:
Test no longer relevant

DanShort12 committed:
Improved docstring for sample_source

shimwell commented:
Congrats on the recent OpenMC merge 
https://github.com/openmc-dev/openmc/pull/1623

DanShort12 committed:
Merge pull request #3 from DanShort12/implement_ci

Implement CI

Implements a CI pipeline to build, test, and publish wheels as artifacts.
Implements a baseline set of tests.

DanShort12 committed:
Merge branch 'master' into serialise_plasma_source

DanShort12 committed:
Remove pugixml as we don't output to XML any more

DanShort12 committed:
Add virtual env to gitignore

DanShort12 committed:
Update implementation to align with openmc

Use std::string everywhere.
Fix usage of unique_ptr --> std::unique_ptr.
sample_source --> source.

DanShort12 committed:
Build and distribute source_sampling

Requires clone and build of OpenMC in GitHub action.
Add source_sampling path to __init__.
Build source_sampling in action and include in wheel.

DanShort12 committed:
Fix typo in package name

DanShort12 committed:
Add black and flake8 to development requirements

DanShort12 committed:
Implement tests of serialisation

DanShort12 committed:
Write a simple example

DanShort12 committed:
Fix flake8 messages

DanShort12 committed:
Run black to reformat files

DanShort12 committed:
Exclude outputs and vscode

DanShort12 committed:
Add tests for sampling functions

DanShort12 committed:
Naming and formatting

Change sample_source -> sample to align with openmc.
Tidy up some function definitions for readability.
Remove unused XML constant.

DanShort12 committed:
Implement public getters for attributes

DanShort12 committed:
Fix tests

pytest.approx syntax was wrong (left over from removing numpy isclose).
Length units should match the input (cm not m).
Move fixure to conftest.py so that it can be shared.

DanShort12 committed:
Small format update

DanShort12 committed:
Fix some typos

seperatrix -> separatrix
pedistal -> pedestal

DanShort12 committed:
Update README

DanShort12 committed:
Supply length units in m

Reinstate conversion to cm when sampling in OpenMC.

DanShort12 committed:
Use openmc model for examples

DanShort12 committed:
Merge pull request #4 from DanShort12/serialise_plasma_source

Serialise plasma source

DanShort12 committed:
Merge branch 'master' into source_sampling_executable

DanShort12 committed:
Align source sampling with parameterised source

Allows the source generator to be run by passing in the parameterised
source.
Also provides an example for how to read the output h5 file.
Requires pugixml to align with OpenMC calls.

DanShort12 commented:
@Shimwell, this PR should now support passing the source parameters into the `source_generator` executable. Hope that's a useful addition!

shimwell commented:
Sorry to be a pain is there any chance this PR can be redirect to the new repository

https://github.com/open-radiation-sources/parametric-plasma-source

@DanShort12

DanShort12 commented:
Superseded by open-radiation-sources/parametric-plasma-source#2
